security/pkg/nodeagent/cache/secretcache.go

```diff
@@ -629,9 +638,16 @@ func (sc *SecretCache) rotate(updateRootFlag bool) {
 		// Re-generate secret if it's expired.
 		if sc.shouldRotate(&secret) {
 			atomic.AddUint64(&sc.secretChangedCount, 1)
-			// Send the notification to close the stream if token is expired, so that client could re-connect with a new token.
-			if sc.isTokenExpired(&secret) {
-				cacheLog.Debugf("%s token expired", logPrefix)
+			if UseRotationWithCert {
+				err := sc.fetcher.CaClient.Reconnect(true)
```

Comment:
I feel like we shouldn't need to reconnect every time. Today, we support using cert to rotate, and we don't do this. So i would expect we do not need to for this as well? *maybe* the first time, but surely not the Nth time? Not sure how it works today, will take a look

Reply:
Specifically https://github.com/istio/istio/blob/65bd2ac632cbd6ec8e3fd85bb9874f685aea0cfd/security/pkg/nodeagent/caclient/providers/citadel/client.go#L137

Reply:
Actually We need to reconnect. Because when the connection is created at first time it will use the existing client cert exists provided in the GetClientCertificate field to authenticate the CSR request. And since that the connection will not extract the cert to authenticate again until we close it and rebuild the connection. So it needs to reconnect, because the cert in the OUTPUT_CERTS has already changed

Reply:
Will the connection closed after the cert expired?

Reply:
yes of course. because the connection is closed